devito/builtins.py

```diff
@@ -53,6 +100,212 @@ def smooth(f, g, axis=None):
         dv.Operator(dv.Eq(f, g.avg(dims=axis)), name='smoother')()
 
 
+def gaussian_smooth(f, sigma=1, _order=4, mode='reflect'):
+    """
+    Gaussian smooth function.
+    """
+    class ObjectiveDomain(dv.SubDomain):
+
+        name = 'objective_domain'
+
+        def __init__(self, lw):
+            super(ObjectiveDomain, self).__init__()
+            self.lw = lw
+
+        def define(self, dimensions):
+            return {d: ('middle', self.lw, self.lw) for d in dimensions}
+
+    def fset(f, g):
+        indices = [slice(lw, -lw, 1) for _ in g.grid.dimensions]
+        slices = (slice(None, None, 1), )*len(g.grid.dimensions)
+        if isinstance(f, np.ndarray):
+            f[slices] = g.data[tuple(indices)]
+        elif isinstance(f, dv.Function):
+            f.data[slices] = g.data[tuple(indices)]
+        else:
+            raise NotImplementedError
+
+    lw = int(_order*sigma + 0.5)
+
+    # Create the padded grid:
+    objective_domain = ObjectiveDomain(lw)
+    try:
+        shape_padded = np.array(f.grid.shape) + 2*lw
+    except AttributeError:
+        shape_padded = np.array(f.shape) + 2*lw
+    grid = dv.Grid(shape=shape_padded, subdomains=objective_domain)
+
+    f_c = dv.Function(name='f_c', grid=grid, space_order=2*lw,
+                      coefficients='symbolic', dtype=np.int32)
+    f_o = dv.Function(name='f_o', grid=grid, coefficients='symbolic', dtype=np.int32)
+
+    weights = np.exp(-0.5/sigma**2*(np.linspace(-lw, lw, 2*lw+1))**2)
+    weights = weights/weights.sum()
+
+    mapper = {}
+    for d in f_c.dimensions:
+        lhs = []
+        rhs = []
+        options = []
+
+        lhs.append(f_o)
+        rhs.append(dv.generic_derivative(f_c, d, 2*lw, 1))
+        coeffs = dv.Coefficient(1, f_c, d, weights)
+        options.append({'coefficients': dv.Substitutions(coeffs),
+                        'subdomain': grid.subdomains['objective_domain']})
+        lhs.append(f_c)
+        rhs.append(f_o)
+        options.append({'subdomain': grid.subdomains['objective_domain']})
+
+        mapper[d] = {'lhs': lhs, 'rhs': rhs, 'options': options}
+
+    initialize_function(f_c, f.data[:], lw,
+                        mapper=mapper,
+                        mode='reflect', name='smooth')
+
+    fset(f, f_c)
+
+    return f
+
+
+def initialize_function(function, data, nbl, mapper=dict(), mode='constant',
+                        name='padfunc'):
+    """
+    Initialize a `Function` with the given ``data``. ``data``
+    does *not* include the ``nbl`` outer/boundary layers; these are added via padding
+    by this function.
+
+    Parameters
+    ----------
+    function : Function
+        The initialised object.
+    data : ndarray of Function
+        The data used for initialisation.
+    nbl : int
+        Number of outer layers (such as PML layers for boundary damping).
+    mapper : dict, optional
+        Dictionary containing, for each dimension of function, a sub-dictionary
```

Comment:
here you probably mean \`function\` as you're referring to a specific variable name